koku/koku/middleware.py

```diff
@@ -161,38 +184,87 @@ def process_request(self, request):
 
-    def get_tenant(self, model, hostname, request):
-        """Override the tenant selection logic."""
-        schema_name = "public"
+    def _check_user_has_access(self, request):
+        """Check if the user has access to Cost Management.
+
+        Args:
+            request (HttpRequest): The incoming request object.
+
+        Raises:
+            PermissionDenied: If the user does not have permissions for Cost Management.
+
+        """
+
+        username = request.user.username
+        if not request.user.admin and request.user.access is None:
+            LOG.warning("User %s does not have permissions for Cost Management.", username)
+            # For /user-access we do not want to raise the exception since the API will
+            # return a false boolean response that the platfrom frontend code is expecting.
+            if request.path != reverse("user-access"):
+                raise PermissionDenied()
+
+    def _get_or_create_tenant(self, request):
+        """Get or create tenant based on the user's schema.
+
+        Args:
+            request(HttpRquest): The incoming request object.
+
+        Returns:
+            Tenant: The tenant object.
+
+        """
+
+        if tenant := self._get_tenant_from_tenant_cache(request):
+            return tenant
+
+        tenant_username = request.user.username
+        schema_name = request.user.customer.schema_name if not is_no_auth(request) else "public"
+        if tenant := self._get_tenant_from_db(tenant_username, schema_name):
+            return tenant
+
+        return self._create_tenant()
+
+    def _get_tenant_from_tenant_cache(self, request):
+        """Get tenant from tenant cache."""
+
         tenant_username = request.user.username
         tenant = KokuTenantMiddleware.tenant_cache.get(tenant_username)
-        if not tenant:
-            if not is_no_auth(request):
-                schema_name = request.user.customer.schema_name
+        return tenant
 
-            tenant = model.objects.filter(schema_name=schema_name).first()
-            if tenant and schema_name != "public":
+    def _get_tenant_from_db(self, tenant_username, schema_name):
+        """Get tenant with the schema from the database."""
+
+        try:
+            tenant = Tenant.objects.get(schema_name=schema_name)
+            if schema_name != "public":
                 with KokuTenantMiddleware.tenant_lock:
                     KokuTenantMiddleware.tenant_cache[tenant_username] = tenant
                     LOG.debug(f"Tenant added to cache: {tenant_username}")
-            elif not tenant:
-                tenant, __ = model.objects.get_or_create(schema_name="public")
+            return tenant
+        except Tenant.DoesNotExist:
+            LOG.info(f"Tenant does not exist: {tenant_username}")
+            return
+
+    def _create_tenant(self):
+        """Create tenant"""
 
+        tenant, __ = Tenant.objects.get_or_create(schema_name="public")
```

Comment:
I know this is what the original code did, but why is the schema always `public`? Just want to make sure this is correct.

Reply:
> Even though the second returned value isn't used, [it should have a descriptive name](https://docs.quantifiedcode.com/python-anti-patterns/maintainability/using_single_letter_as_variable_name.html) for clarity.

Isn't it common practice to put an `_` for an unused variable? Otherwise I think flake8 would stop you on an [F841](https://www.flake8rules.com/rules/F841.html)

Reply:
I think using `__` indicates that the value is intentionally being ignored or not needed in the context of the code as the case here.

Reply:
> Isn't it common practice to put an _ for an unused variable? Otherwise I think flake8 would stop you on an [F841](https://www.flake8rules.com/rules/F841.html)

I'm not sure how common it is in general. It's certainly possible and accepted in certain code bases. For our code I would highly recommend against this since we also have adopted the conversion of aliasing variations of `gettext` to `_`, creating the potential for conflicts.

`pyflakes` does not report unused values in tuple unpacking and [has not for a while](https://github.com/PyCQA/pyflakes/commit/4417b7730c3e42087d3adb56d13bd4756d05b0bc).

Reply:
I was curious about the `gettext` alias to `_`. Turns out that's [documented](https://docs.python.org/3/library/gettext.html#gettext.gettext).

Reply:
This is a side topic for sure. Just sharing some thoughts about other things I found.

Some interesting notes in the [Django docs](https://docs.djangoproject.com/en/4.2/topics/i18n/translation/#standard-translation) about this.

> It’s convention to import this as a shorter alias, _, to save typing.

>  In Django, we have chosen not to follow this practice, for a couple of reasons

I find aliasing a function to `_` suuuuper confusing. Even though it's a documented in the `gettext` docs and `gettext.install()` puts `_()` in the , the extra typing is a small price to pay for the clarity it brings.